[Permission] Rework

lukmccall opened the pull request:
# Why

Resolves #4972.
Resolves #5831.
Resolves #4289.
Resolves #4618.

# How

- iOS
  - [x] general clean up. 
  - [x] move logic responsible for asking/getting more than one permission to js side.
  - [x] extract base class for requesters.
  - [x] move each requester to separate podspec.  
  - [x] remove `EXPermissionsModule protocol.
  - [x] make requesters registrable. 
  - [x] add `[_constantsBinding.appOwnership isEqualToString:@"expo"]` to `EXScopedPermissions`. ([see](https://github.com/expo/expo/commit/7b5cc3cff209c57f01e4ede1518fea946d949f36))
  - [x] modules, which need permissions should have own native method to handle own permission.
    - [x] move requesters to modules 
- Android
  - [x] rewrite to kotlin.
  - [x] general clean up.
  - [x] move logic responsible for `scoped permission ask` to activity.
  - [x] remove `PermissionsManager`.
  - [x] remove `PermissionsServiceBinding`. 
  - [x] clean `PermissionsModule` class.
  - [x] do something with `systemBrightness`.
  - [x] cache `mPermissionsAskedFor`. 
  - [x] modules, which need permissions should have own native method to handle own permission.
    - [x] move logic from `PermissionsModule` to `PermissionsService` 
    - [x] rewrite native api for permissions - make it similar to iOS 
    - [x] use List<String> insted of String[] in native interface
    - [x] move requsters to modules
  - [x] resolve #5831.
- Both
  - [x] resolve #4289.
  - [x] resolve #4618 (not fixed in expo client).

# Flowgraph

- iOS
![iOS_flow_graph](https://user-images.githubusercontent.com/9578601/62854204-e6af5780-bcee-11e9-81cf-18e7ade1763f.png)

- Android
![android_flow_graph](https://user-images.githubusercontent.com/9578601/68225468-c0990b00-fff0-11e9-9d9d-5f3834c62c4d.png)

# Test Plan

- test-suite passed.
- I tested it with a simple [demo](https://snack.expo.io/@lukaszkosmaty/permissions-refactoring) (run on several experiences to be able to test scoped permissions).

## Test cases:
| Test | Android | iOS |
|-----|:---:|:---:|
| Permissions are properly set in settings. | 🆗 | 🆗 |
| Permissions are properly scoped (switching between two apps). | 🆗 | 🆗 |
| Global combined status is correct. | 🆗 | 🆗 |
| Scoped status is correct (the previous test but enforce using scoped permission). | 🆗| 🆗|
| Check if `canAskAgain` and `granted` fields have the correct value. | 🆗| 🆗|
| Test added module methods (getting/asking using, for example, the `Camera` module). | 🆗| 🆗|
| Test backward compatibility. | 🆗 | 🆗|
| Changing permissions in system settings. | 🆗| 🆗| 
| Bare workflow. | 🆗 | 🆗 |

lukmccall committed:
[Permissions] Extract base class for requester

lukmccall committed:
[Permissions] Refactor

lukmccall committed:
[Permissions] Make podspecs for each requesters

lukmccall committed:
[Permissions] Add missing self ensurement to requesters

lukmccall committed:
[Permissions] Android fix

lukmccall committed:
[Permissions] Rewrite modul to kotlin

lukmccall commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/PermissionsModule.kt:
```diff
@@ -0,0 +1,406 @@
+   * Throws IllegalStateException there's no Permissions module present.
+   */
+  private fun arePermissionsGranted(permissions: Array<String>): Boolean {
+    if (mPermissions != null) {
```
Here should be `mPermissions == null`. I will fix it later.

lukmccall committed:
[Permissions] Move scoped ask to activity

lukmccall committed:
[Permissions] Remove permissions manager

lukmccall committed:
[Permissions] Remove PermissionsServiceBinding

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/utils/PermissionsHelper.java:
```diff
@@ -56,7 +56,7 @@ public boolean requestPermissions(Exponent.PermissionsListener listener, String[
     for (String permission : permissions) {
       if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
-          ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
+          /* ask for global permission */ activity.checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
```
```suggestion
          /* check for global permission */ activity.checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
```

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/experience/DetachedModuleRegistryAdapter.java:
```diff
@@ -42,7 +42,7 @@ public DetachedModuleRegistryAdapter(ReactModuleRegistryProvider moduleRegistryP
 
     // Overriding ScopedUIManagerModuleWrapper from ReactAdapterPackage
-    moduleRegistry.registerInternalModule(new ScopedUIManagerModuleWrapper(reactContext, experienceId, manifest.optString(ExponentManifest.MANIFEST_NAME_KEY)));
+    moduleRegistry.registerInternalModule(new ScopedUIManagerModuleWrapper(reactContext, experienceId, manifest.optString(ExponentManifest.MANIFEST_NAME_KEY))); // todo: I think I can delete this line but I'm not sure :\
```
Find out then 😜 cc @tsapeta @sjchmiela

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/experience/BaseExperienceActivity.java:
```diff
@@ -263,8 +284,55 @@ protected void registerForNotifications() {
+      mPermissionsHelper = null;
+    }
+  }
+
```
```suggestion
```

bbarthec commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.m:
```diff
@@ -40,190 +33,125 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
     return YES;
   }
-
+  
```
```suggestion

```

bbarthec commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.m:
```diff
@@ -40,190 +33,125 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
     return @"photos";
   }
-
+  
```
```suggestion

```

lukmccall committed:
[Permissions] Make requesters

lukmccall committed:
[Permissions] Improve code styling

lukmccall committed:
[Permissions] Make SystemBrightnessRequester (#5147)

* Make SystemBrightnessRequester

* Make type enum

Co-Authored-By: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/CalendarRequester.kt:
```diff
@@ -0,0 +1,42 @@
+      putString(EXPIRES_KEY, PERMISSION_EXPIRES_NEVER)
+    }
+  }
+}
```
newline needed

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/CameraRollRequester.kt:
```diff
@@ -0,0 +1,42 @@
+      putString(EXPIRES_KEY, PERMISSION_EXPIRES_NEVER)
+    }
+  }
+}
```
newline needed

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/LocationRequester.kt:
```diff
@@ -0,0 +1,50 @@
+      putBundle("android", Bundle().apply { putString("scope", scope) })
+    }
+  }
+}
```
newline needed

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/NotificationRequester.kt:
```diff
@@ -0,0 +1,22 @@
+      putString(EXPIRES_KEY, PERMISSION_EXPIRES_NEVER)
+    }
+  }
+}
```
newline needed

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/PermissionRequester.kt:
```diff
@@ -0,0 +1,60 @@
+  fun getPermission(): Bundle
+
+  fun getPermissionToAsk(): Array<String>
+}
```
newline needed

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/PermissionRequester.kt:
```diff
@@ -0,0 +1,60 @@
+              putString(STATUS_KEY, GRANTED_VALUE)
+            }
+            else -> {
+              putString(STATUS_KEY, UNDETERMINED_VALUE)
```
lift up `putString`

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/RemindersRequester.kt:
```diff
@@ -0,0 +1,19 @@
+      putString(EXPIRES_KEY, PERMISSION_EXPIRES_NEVER)
+    }
+  }
+}
```
newline

bbarthec commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/SimpleRequester.kt:
```diff
@@ -0,0 +1,16 @@
+  override fun getPermission(): Bundle {
+    return PermissionRequester.getSimplePermission(permission)
+  }
+}
```
newline

bbarthec commented on packages/expo-permissions/ios/EXPermissions/EXPermissionBaseRequester.m:
```diff
@@ -0,0 +1,16 @@
+@implementation EXPermissionBaseRequester
+
+- (void)requestPermissionsWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject {
+  // todo: change it to something better
```
ping

bbarthec commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -71,109 +125,51 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
-                      withRejecter:reject];
+  [self askForPermissionWithType:permissionType
+                     withResults:resolve 
+                     withRejecter:reject];
```
```suggestion
                    withRejecter:reject];
```

bbarthec commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -182,83 +178,64 @@ - (BOOL)hasGrantedPermission:(NSString *)permissionType
-                   withResolver:globalPermissionResolver
+ 
+  [self askForGlobalPermission:permissionType
+                   withResolver:onResults
```
```suggestion
                  withResolver:onResults
```

bbarthec commented on packages/expo-permissions/src/Permissions.ts:
```diff
@@ -51,4 +88,4 @@ async function _handlePermissionsRequestAsync(
     permissions,
   };
-}
+}
```
newline

lukmccall committed:
[Permissions] Rewrite android native api (#5186)

* Move logic form module to service

* Remove static var from PermissionsService

* Make getPermission not throwable

* Store asked permissions

lukmccall committed:
[Permissions] Make requester registrable (#5187)

* Clean iOS

* Fix scoped permissions status (never undetermind)

* Requesters belong to EXPermission

* Remove permissions delegate

* Make requester registrable

lukmccall committed:
[Permissions] Small fixes

bbarthec commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -28,266 +29,256 @@ @implementation EXPermissions
-- (void)askForPermissions:(NSArray<NSString *> *)permissionsTypes
-              withResults:(void (^)(NSArray<NSDictionary *> *))onResults
-             withRejecter:(UMPromiseRejectBlock)reject
+- (void)askForPermissionUsingRequesterClass:(Class)requesterClass
```
This method will handle requesting permission once requesters would be moved to separate modules, right?

bbarthec commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -28,266 +29,256 @@ @implementation EXPermissions
+    return reject(@"E_PERMISSIONS_UNKNOWN", [NSString stringWithFormat:@"Unrecognized requester: %@", NSStringFromClass(requesterClass)], nil);
+  }
+  
+  BOOL isGranted = [EXPermissions statusForPermission:permission] == UMPermissionStatusGranted;
```
What about permissions returning composite results that not necessarily are explicitly granted?
E.g. `NOTIFICATIONS` right now comes with this scenario.

bbarthec commented on packages/unimodules-permissions-interface/ios/UMPermissionsInterface/UMPermissionsInterface.h:
```diff
@@ -1,18 +1,43 @@
 - (BOOL)hasGrantedPermission:(NSString *)permissionType;
 
 - (void)askForPermission:(NSString *)permissionType
               withResult:(void (^)(NSDictionary *))onResult
```
I wonder whether this actually cannot be
```suggestion
              withResult:(UMPromiseResolveBlock)onResult
```
❓

bbarthec commented on packages/unimodules-permissions-interface/ios/UMPermissionsInterface/UMPermissionsInterface.h:
```diff
@@ -1,18 +1,43 @@
+- (BOOL)hasGrantedPermissionUsingRequesterClass:(Class)requesterClass;
+
+- (void)askForPermissionUsingRequesterClass:(Class)requesterClass
+              withResult:(void (^)(NSDictionary *))onResult
```
Same here:
```suggestion
              withResult:(UMPromiseResolveBlock)onResult
```
❓ 

If so, please adjust all the occurrences 😉

bbarthec reviewed (commented):
LGTM 💯 Amazing work 👍
One suggestion:
Maybe these's some clever way to rewrite
```obj-c
%{
	@"status": sth,
	@"expires": sth
}
```

into some POJO-style or NSDictionary-based?

There are many places where this construction is used and it's always typed by hand, but maybe we can somehow enforce more class-based behaviours instead of JSON-like behaviour.

lukmccall committed:
[Permissions] Add app ownership check and fix code style

lukmccall committed:
[Permissions] Rework native api (#5418)

* [Permissions] Move requesters to modules on iOS

* [Permissions] Add bool granted

* [Permissions] Add to native modules, permissions related methods

* [Permissions] Repair CI

* [Permissions] Fix typo

* [Permissions] Reduce boilerplate code

* [Permissions] Fix error description

* [Permissions] Export constants for permissions status

* [Permissions] Rename methods wrapper

lukmccall committed:
[Permissions] Remove scoped permissions from standalon apps

lukmccall committed:
[Permissions] Fix lateinit and scoped permissions

Location module in OnHostResume calls getPermissions method, before Activity Provider is available.

lukmccall committed:
[Permissions] Add `never ask again` support

lukmccall committed:
[Permissions] Scoping only dangerous permission

bbarthec reviewed (commented):
Great job! 💯 Let's merge it once everything is resolved and proceed to testing 😉 
One global comment: think more about variables/methods/functions names to provide as much information as possible 👍 
One request: maybe it's possible to draw some diagram showing what and how is happening during PermissionRequester registration and asking for/getting permission starting from JS side 🤔 🙇 🙏 (If I was to ask I would do sth like [this](https://developer.android.com/guide/components/activities/activity-lifecycle#alc) using draw.io) 😉

lukmccall committed:
[Permissions] Rename  `neverAskAgain` to `canAskAgain`

lukmccall commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.h:
```diff
@@ -6,38 +6,16 @@
-- (NSDictionary *)getPermissionsForResource:(NSString *)resource;
-
-+ (NSString *)permissionStringForStatus:(EXPermissionStatus)status;
+- (void)askForGlobalPermissionUsingRequesterClass:(Class)requesterClass
```
Yes that's true

lukmccall commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/PermissionsService.kt:
```diff
@@ -0,0 +1,260 @@
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider::class.java)
+        ?: throw IllegalStateException("Couldn't find implementation for ActivityProvider.")
+    moduleRegistry.getModule(UIManager::class.java).registerLifecycleEventListener(this)
+    mPermissionsAskedStorage = context.applicationContext.getSharedPreferences(PREFERENCE_FILENAME, Context.MODE_PRIVATE)
```
I've renamed it.

lukmccall commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/PermissionsService.kt:
```diff
@@ -0,0 +1,260 @@
+typealias PermissionsListener = (result: IntArray) -> Unit
+
+class PermissionsService(val context: Context) : InternalModule, Permissions, LifecycleEventListener {
+  private var mActivityProvider: ActivityProvider? = null
```
Location module in OnHostResume calls `getPermissions` method, before Activity Provider is available. So, that's why I've decided to remove lateinit.

bbarthec commented on packages/expo-permissions/src/Permissions.ts:
```diff
@@ -1,5 +1,7 @@
-import { coalesceExpirations, coalesceStatuses } from './CoalescedPermissions';
+import { coalesceExpirations, coalesceStatuses, coalesceCanAskAgin } from './CoalescedPermissions';
```
```suggestion
import { coalesceExpirations, coalesceStatuses, coalesceCanAskAgain } from './CoalescedPermissions';
```

bbarthec reviewed (approved):
I've found one small typo 😛 
Fix, merge master resolving conflicts, ensure CI builds correctly and merge 💯 
Let's ship it! 😄

lukmccall committed:
[Permissions] Run pod install

bbarthec reviewed (approved):
From what I can see:
- `packages/react-native-unimodule` shouldn't probably be updated -> it's submodule, so please `git submodule update` after rebasing on/merging master
- `yaga` pod shouldn't be deleted 🤔 Please revert this (`et ios-generate-dynamic-macros` after rebasing)

Please fix and merge 🎉

lukmccall committed:
[Permissions] Update docs

lukmccall committed:
[Permissions] Fix

lukmccall committed:
[Permissions] Fix CI

lukmccall committed:
[Permissions] Fix linter warnings

brentvatne commented on docs/pages/versions/unversioned/sdk/imagepicker.md:
```diff
@@ -18,6 +18,30 @@ import SnackEmbed from '~/components/plugins/SnackEmbed';
+
+A promise that resolves to an object of type [PermissionsResponse](#permissionsresponse).
+
+### `ImagePicker.getCamerPermissionsAsync()`
```
```suggestion
### `ImagePicker.getCameraPermissionsAsync()`
```

brentvatne commented on docs/pages/versions/unversioned/sdk/imagepicker.md:
```diff
@@ -18,6 +18,30 @@ import SnackEmbed from '~/components/plugins/SnackEmbed';
+
+Checks user's permissions for accessing camera. Alias for `Permissions.getAsync(Permissions.CAMERA)`.
+
+### `ImagePicker.getCamerRollPermissionsAsync()`
```
```suggestion
### `ImagePicker.getCameraRollPermissionsAsync()`
```

lukmccall committed:
[Permissions] Fix typo and imports after rebase

brentvatne reviewed (commented):
a couple questions re: documentation

1. how will we need to update the [react-native-unimodules readme](https://github.com/unimodules/react-native-unimodules)?
2. how will we need to update the installation / configuration instructions in the READMEs for each related module? eg: https://github.com/expo/expo/blob/master/packages/expo-calendar/README.md

lukmccall committed:
[Permissions] Small improvements in Android

tsapeta commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.h:
```diff
@@ -9,7 +9,9 @@ NS_ASSUME_NONNULL_BEGIN
-- (EXPermissionStatus)getPermission:(NSString *)permissionType forExperience:(NSString *)experienceId;
+- (UMPermissionStatus)getPermission:(NSString *)permissionType forExperience:(NSString *)experienceId;
+// TODO: Remove once SDK32 is phased out
+- (BOOL)hasGrantedPermission:(NSString *)permission forExperience:(NSString *)experienceId; // used in SDKs 29–32
```
Can we remove it? 🤔

tsapeta commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.m:
```diff
@@ -40,185 +32,123 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
-  // not in Expo Client - invoke allow action for each permission type
-  if (![_constantsBinding.appOwnership isEqualToString:@"expo"]) {
-    NSMutableDictionary *results = [NSMutableDictionary new];
+  NSDictionary* globalPermissions = [super getPermissionUsingRequesterClass:requesterClass];
```
```suggestion
  NSDictionary *globalPermissions = [super getPermissionUsingRequesterClass:requesterClass];
```

tsapeta commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.m:
```diff
@@ -40,185 +32,123 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
-  if (![_constantsBinding.appOwnership isEqualToString:@"expo"]) {
-    NSMutableDictionary *results = [NSMutableDictionary new];
+  NSDictionary* globalPermissions = [super getPermissionUsingRequesterClass:requesterClass];
+  NSString* permissionType = [requesterClass permissionType];
```
```suggestion
  NSString *permissionType = [requesterClass permissionType];
```

tsapeta commented on packages/expo-brightness/ios/EXBrightness/EXSystemBrightnessRequester.m:
```diff
@@ -0,0 +1,23 @@
+
+#import <EXBrightness/EXSystemBrightnessRequester.h>
+
+@implementation EXSystemBrightnessRequester
```
I think we should call these requesters as permission requesters, so let's change it to `EXSystemBrightnessPermissionRequester`

tsapeta commented on packages/expo-brightness/src/Brightness.ts:
```diff
@@ -8,6 +8,19 @@ export enum BrightnessMode {
+  granted: boolean;
+  canAskAgain: boolean;
+};
+
```
It might be better to have these types defined in `unimodules-permissions-interface` and re-export them in modules using them.

tsapeta commented on packages/expo-calendar/ios/EXCalendar/EXCalendarRequester.h:
```diff
@@ -0,0 +1,7 @@
+
+#import <UMPermissionsInterface/UMPermissionsInterface.h>
+
+@interface EXCalendarRequester : NSObject<UMPermissionsRequester>
```
```suggestion
@interface EXCalendarPermissionRequester : NSObject<UMPermissionsRequester>
```

tsapeta commented on packages/expo-calendar/ios/EXCalendar/EXCalendarRequester.m:
```diff
@@ -1,18 +1,17 @@
-#import <EXPermissions/EXCalendarRequester.h>
+#import <EXCalendar/EXCalendarRequester.h>
```
```suggestion
#import <EXCalendar/EXCalendarPermissionRequester.h>
```

tsapeta commented on packages/expo-calendar/ios/EXCalendar/EXRemindersRequester.h:
```diff
@@ -0,0 +1,7 @@
+
+#import <UMPermissionsInterface/UMPermissionsInterface.h>
+
+@interface EXRemindersRequester : NSObject<UMPermissionsRequester>
```
```suggestion
@interface EXRemindersPermissionRequester : NSObject<UMPermissionsRequester>
```

tsapeta commented on packages/expo-calendar/ios/EXCalendar/EXRemindersRequester.m:
```diff
@@ -1,17 +1,17 @@
-#import <EXPermissions/EXRemindersRequester.h>
+#import <EXCalendar/EXRemindersRequester.h>
```
```suggestion
#import <EXCalendar/EXRemindersPermissionRequester.h>
```

tsapeta commented on packages/expo-camera/ios/EXCamera/EXCamera.m:
```diff
@@ -4,6 +4,7 @@
 #import <EXCamera/EXCamera.h>
 #import <EXCamera/EXCameraUtils.h>
 #import <EXCamera/EXCameraManager.h>
+#import <EXCamera/EXCameraRequester.h>
```
```suggestion
#import <EXCamera/EXCameraPermissionRequester.h>
```

tsapeta commented on packages/expo-camera/ios/EXCamera/EXCamera.m:
```diff
@@ -578,8 +579,7 @@ - (void)startSession
 #endif
-  NSDictionary *cameraPermissions = [_permissionsManager getPermissionsForResource:@"camera"];
-  if (![cameraPermissions[@"status"] isEqualToString:@"granted"]) {
+  if (![_permissionsManager hasGrantedPermissionUsingRequesterClass:[EXCameraRequester class]]) {
```
```suggestion
  if (![_permissionsManager hasGrantedPermissionUsingRequesterClass:[EXCameraPermissionRequester class]]) {
```

tsapeta commented on packages/expo-camera/ios/EXCamera/EXCameraManager.m:
```diff
@@ -1,16 +1,19 @@
 #import <EXCamera/EXCamera.h>
 #import <EXCamera/EXCameraManager.h>
 #import <EXCamera/EXCameraUtils.h>
+#import <EXCamera/EXCameraRequester.h>
```
```suggestion
#import <EXCamera/EXCameraPermissionRequester.h>
```

tsapeta commented on docs/pages/versions/unversioned/sdk/audio.md:
```diff
@@ -18,6 +18,24 @@ For [managed](../../introduction/managed-vs-bare/#managed-workflow) apps, you'll
+
+#### Returns
+
+A promise that resolves to an object of type [PermissionsResponse](#permissionsresponse).
```
I'm wondering if it isn't better to document `PermissionsResponse` type in Permissions docs and just link to that page here 🤔

tsapeta commented on packages/expo-av/ios/EXAV/EXAV.m:
```diff
@@ -7,10 +7,12 @@
 #import <EXAV/EXAV.h>
 #import <EXAV/EXAVPlayerData.h>
 #import <EXAV/EXVideoView.h>
+#import <EXAV/EXAudioRecordingRequester.h>
```
```suggestion
#import <EXAV/EXAudioRecordingPermissionRequester.h>
```

tsapeta reviewed (changes requested):
Generally looks good, but I'd make a few (but wide) changes:
- Rename `*Requester` classes to `*PermissionRequester` - it sounds better and is more descriptive about what the requester actually does.
- Move `PermissionsStatus` and `PermissionsResponse` types to `unimodules-permissions-interface` and their docs to Permissions docs page. Having them defined in every unimodule is not a good idea because we will have to maintain one, the same type, in several places.
These types cannot be in ze `expo-permissions` because we don't want other unimodules to depend on it. Interface seems to be a good alternative and it's going to be a dependency of all unimodules with permission requesters. (Also make sure the interface is actually a dependency of them).

bbarthec commented:
Just to link this rework with web changes: https://github.com/expo/expo/pull/6115

lukmccall committed:
[Permissions] Make test-suits pass

lukmccall commented on packages/expo-barcode-scanner/build/BarCodeScanner.d.ts:
```diff
@@ -10,6 +10,16 @@ export declare type BarCodeEventCallbackArguments = {
+export declare type PermissionsResponse = {
+    status: typeof PermissionsStatus[keyof typeof PermissionsStatus];
+    expires: 'never' | number;
+    granted: boolean;
```
Add canAskAgain

lukmccall commented on packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java:
```diff
@@ -748,6 +743,35 @@ private void stopWatching() {
+    } else if (accessFineLocation.getStatus() == PermissionsStatus.DENIED && accessCoarseLocation.getStatus() == PermissionsStatus.DENIED) {
+      status = PermissionsStatus.DENIED.getStatus();
+    }
+
```
Add granted key.

esamelson commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -28,266 +27,209 @@ @implementation EXPermissions
+
+- (void)getPermissionUsingRequesterClass:(Class)requesterClass
+                              withResult:(UMPromiseResolveBlock)onResult
+                            withRejecter:(UMPromiseRejectBlock)reject
```
```suggestion
                              resolver:(UMPromiseResolveBlock)resolve
                           rejecter:(UMPromiseRejectBlock)reject
```
these are the conventions in the rest of the expo modules, not sure why they were different in this one but let's stick with the common conventions.

esamelson commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -28,266 +27,209 @@ @implementation EXPermissions
-            withRejecter:(UMPromiseRejectBlock)reject
+- (void)askForPermissionUsingRequesterClass:(Class)requesterClass
+                                 withResult:(UMPromiseResolveBlock)onResult
+                               withRejecter:(UMPromiseRejectBlock)reject
```
same comment here as above -- should be resolver:resolve and rejecter:reject

esamelson commented on packages/expo-permissions/ios/EXPermissions/EXPermissions.m:
```diff
@@ -28,266 +27,209 @@ @implementation EXPermissions
+   
+- (void)askForGlobalPermissionUsingRequesterClass:(Class)requesterClass
+                                     withResolver:(UMPromiseResolveBlock)resolver
+                                     withRejecter:(UMPromiseRejectBlock)reject
```
same comment here, and in the rest of the methods added to this module -- should be resolver:resolve and rejecter:reject

esamelson commented on packages/expo-permissions/src/Permissions.ts:
```diff
@@ -30,13 +32,47 @@ export const REMINDERS = 'reminders';
+
+  let permissions = {};
+  for (let type of types) {
+    permissions[type] = await _handleSinglePermissionRequestIOSAsync(type, handlePermission);
```
what is the reason for moving the multi-permissions requests on iOS to the JS side? this is less efficient because it requires asynchronous calls over the bridge for each permission in the request rather than just one for the entire request.

lukmccall replied:
I moved it to JS side because I thought, it'll be easier to maintain this in the future. Native iOS API wasn't created to request for multiple permissions at once. That's why our code was a little bit messy and difficult to read. 
Moreover, asking/getting `permissions` isn't something that user doing very often. Mostly, it's happed once per `permission`. So I think, it wouldn't affect performance a lot.

esamelson replied:
OK, that makes sense. Thanks for explaining.

esamelson commented on packages/unimodules-permissions-interface/ios/UMPermissionsInterface/UMPermissionsInterface.h:
```diff
@@ -1,18 +1,37 @@
-- (BOOL)hasGrantedPermission:(NSString *)permissionType;
+- (void)registerRequesters:(NSArray<id<UMPermissionsRequester>> *)newRequesters;
+
+- (void)getPermissionUsingRequesterClass:(Class)requesterClass
```
same comment as above for **both** methods added here -- should be resolver:resolve and rejecter:reject

esamelson commented on packages/unimodules-permissions-interface/ios/UMPermissionsInterface/UMPermissionsMethodsDelegate.h:
```diff
@@ -0,0 +1,22 @@
+
++ (void)getPermissionWithPermissionsManager:(id<UMPermissionsInterface>)permissionsManager
+                              withRequester:(Class)requesterClass
+                                 withResult:(UMPromiseResolveBlock)resolve
```
same comment for both of these as well -- use `resolver:resolve` and `rejecter:reject`

esamelson reviewed (changes requested):
Hi there! Thanks for taking on this job 🙂 It looks quite good overall! I've left a number of comments -- many are small style/spelling nits, but there are some important logic-related comments too.

In a big PR like this, having a thorough test plan is really important. Here are a few things I think need to be tested before we can land this:
- For standalone iOS and Android apps -- upgrading a current SDK 35 standalone app to a new build with this PR's changes, and making sure the previously granted permissions are still granted and do not need to be requested again. Since we're modifying the scoped permissions implementation and don't want this to break in people's production apps, this is critical to test.
- Same as above, but in the iOS and Android Expo clients. This is less critical than in production standalone apps, but still important for developer trust.
- Also, should test for the expected behavior when permissions are granted or denied from the initial request, then the status is subsequently changed in OS settings manually. This is difficult to write test-suite tests for but an important part of the API.

^If you've already tested any of those, awesome, just wanted to be sure! 😄 

One aside -- in the future please split giant PRs like this into smaller components (e.g. one PR for each unimodule). This makes it much more ergonomic and efficient to review.

lukmccall commented on android/expoview/src/main/java/host/exp/exponent/utils/PermissionsHelper.java:
```diff
@@ -1,10 +1,14 @@
 package host.exp.exponent.utils;
```
It's for ensuring back compatibility. We'll remove it later.

esamelson replied:
OK. Could we add a big comment to this file explaining that, as well as any places it's imported & used?

lukmccall commented on packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/SimpleRequester.kt:
```diff
@@ -0,0 +1,36 @@
+      val permissionsStatus = when {
+        getAndroidPermissions().all { permissionsResponse.getValue(it).status == PermissionsStatus.GRANTED } -> PermissionsStatus.GRANTED
+        getAndroidPermissions().all { permissionsResponse.getValue(it).status == PermissionsStatus.DENIED } -> PermissionsStatus.DENIED
+        else -> PermissionsStatus.UNDETERMINED
```
It isn't combined status. We add combined values in js before returning them to the user.

esamelson replied:
Ah okay, I see. Thanks for the explanation. I still wonder whether this is the correct behavior, or if we should return `denied` if the user has denied any of the individual permissions.

Either way, we should document this -- would be great to have our docs be at this level of detail.

lukmccall committed:
[Permissions] Add requested changes

lukmccall committed:
[Permissions] Move ts types to unimodule

lukmccall commented:
> In a big PR like this, having a thorough test plan is really important. Here are a few things I think need to be tested before we can land this:
> 
> * For standalone iOS and Android apps -- upgrading a current SDK 35 standalone app to a new build with this PR's changes, and making sure the previously granted permissions are still granted and do not need to be requested again. Since we're modifying the scoped permissions implementation and don't want this to break in people's production apps, this is critical to test.
> * Same as above, but in the iOS and Android Expo clients. This is less critical than in production standalone apps, but still important for developer trust.
> * Also, should test for the expected behaviour when permissions are granted or denied from the initial request, then the status is subsequently changed in OS settings manually. This is difficult to write test-suite tests for but an important part of the API.

I've done some testing. Also, I've tied to include all edge cases like changing permissions in setting manually (in expo-client it's inconsistent with settings. However, in the future we could add something like permissions screen to the client, which will display all applications with their permissions and possibility to change them) or changing between apps with different SDK versions.

Moreover, I'm going to test iOS again on Monday. To make sure that everything works fine.

esamelson commented:
Cool, sounds good. It would be awesome if you could be explicit about what cases you've tested -- for example I really like the checklist you made in the issue description detailing the steps, would be great to have something like that for test cases. That makes it easier for us to determine if anything got missed, or know if you're super thorough and have covered everything and beyond!

One question -- what exactly do you mean by 
> in expo-client it's inconsistent with settings
?

lukmccall committed:
[Permissions] Add missing requester

lukmccall committed:
[Permissions] Small fixes

lukmccall committed:
[Permissions] Fix after rebase

bbarthec commented on CHANGELOG.md:
```diff
@@ -29,10 +31,16 @@ This is the log of notable changes to the Expo client that are developer-facing.
 - Fixed `Calendar.getCalendarsAsync` rejecting on iOS 13, when `source.name` is null. ([#5994](https://github.com/expo/expo/pull/5994) by [@lukmccall](https://github.com/lukmccall))
 - Fixed handling URI with no scheme in `ExpoFileSystem`. ([#5904](https://github.com/expo/expo/pull/5904) by [@bbarthec](https://github.com/bbarthec))
 - Fixed `FileSystem#deleteAsync` in older Android SDKs. ([#5923](https://github.com/expo/expo/pull/5923) by [@bbarthec](https://github.com/bbarthec))
+  <<<<<<< HEAD
```
It's not right 😉

bbarthec commented on CHANGELOG.md:
```diff
@@ -29,10 +31,16 @@ This is the log of notable changes to the Expo client that are developer-facing.
+  =======
+- Fixed crashes when the user tries to download a file through `react-native-webview` without granted storage permission on Android. ([#5061](https://github.com/expo/expo/pull/5061) by [@lukmccall](https://github.com/lukmccall))
+- Fixed `Permissions.getAsync` result, which was inconsistent with iOS settings. ([#5061](https://github.com/expo/expo/pull/5061) by [@lukmccall](https://github.com/lukmccall))
+  > > > > > > > [Permissions] Add requested changes
```
Please take care of it 🙏

bbarthec commented on ios/Exponent/Versioned/Core/UniversalModules/Permissions/EXScopedPermissions.m:
```diff
@@ -40,185 +32,123 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+  UM_WEAKIFY(self);
+  [UMUtilities performSynchronouslyOnMainThread:^{
+    UM_ENSURE_STRONGIFY(self);
+    // TODO: below line is sometimes failing with: "Presenting view controllers on detached view controllers is discourage"
```
AFAIK this comment line is dangling here for a while, but I believe changes that are made quite a long time ago already fixed it 🤔

lukmccall replied:
Unfortunately, I couldn't reproduce this behaviour and I've never encountered this error.

lukmccall committed:
[Permissions] Update docs

lukmccall commented on android/expoview/src/main/java/host/exp/exponent/experience/BaseExperienceActivity.java:
```diff
@@ -82,6 +114,14 @@ protected void onCreate(Bundle savedInstanceState) {
 
     NativeModuleDepsProvider.getInstance().inject(BaseExperienceActivity.class, this);
+
+    if (mManifest == null) {
```
I've added this here because `HomeActivity` couldn't set this properly. Unforynelty, I'm not sure how can I set it in a suitable place.

esamelson replied:
This should definitely be set in `HomeActivity`. Is the problem that you aren't able to get the manifest from that activity? If so, does this comment help? https://github.com/expo/expo/pull/6162#discussion_r343451589 😄

lukmccall commented:
An important note here. We need to publish a new home after merge. Without this, iOS will not work (it needs a new js code to work properly).

sjchmiela commented:
Re: https://github.com/expo/expo/pull/5061#issuecomment-549575829 : you can publish home right now without breaking anything!

When building Expo client from `master` it fetches the app from [`dev-home-config.json`](https://github.com/expo/expo/blob/master/dev-home-config.json), which is a file you can modify and the change will only have an effect on your branch.

To publish dev home, use `et publish-dev-expo-home`. It should relogin to `@expo-home-dev` user, publish `home` and save its experience URL to `dev-home-config`. We will all get this change once the PR with this change is merged. 🙂

lukmccall committed:
[Permissions] Publish dev home

esamelson commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/ContactsModule.java:
```diff
@@ -76,18 +76,46 @@ public void onCreate(ModuleRegistry moduleRegistry) {
+      promise.reject("E_NO_PERMISSIONS", "Permissions module is null. Are you sure all the installed Expo modules are properly linked?");
+      return;
+    }
+    if (permissionsManager.isPermissionPresentInManifest(Manifest.permission.WRITE_CALENDAR)) {
```
it looks like this still needs to be fixed? unless github is showing me the wrong thing

lukmccall replied:
Thanks for notice this. I fixed it.

esamelson reviewed (commented):
a few followups to previous comments but this otherwise looks good 👍 awesome work!!

lukmccall committed:
[Permissions] Add kotlin dependency to build.gradle

lukmccall committed:
[Permissions] Fix permissions in ContactsModule

lukmccall committed:
[Permissions] Set manifest in HomeActivity class

lukmccall committed:
[Permissions] Regenerate pods projects
